docs/Custom_Operation_for_GPUs.md

```diff
@@ -0,0 +1,2323 @@
+
+We can teach JAX that `rms_norm_bwd` is the backward operation for `rms_norm_fwd`, using `jax.custom_vjp` and its convention.  As the first step, we need to refine the definition of `rms_norm_fwd` and `rms_norm_bwd`.
+
+```python
+# rms_norm_fwd was previously defined as
+#
+# def rms_norm_fwd(x, weight, eps=1e-05):
+#     output, invvar = _rms_norm_fwd_p.bind(x, weight, eps=eps)
+#     return output
+#
+def rms_norm_fwd(x, weight, eps=1e-05):
+    output, invvar = _rms_norm_fwd_p.bind(x, weight, eps=eps)
+    return output, (invvar, x, weight)
+
+
+# rms_norm_bwd was previously defined as
+#
+# def rms_norm_bwd(g, invvar, x, weight, eps):
+#     grad_input, grad_weight, part_grad = _rms_norm_bwd_p.bind(
+#         g, invvar, x, weight, eps=eps
+#     )
+#     return grad_input, grad_weight
+#
+def rms_norm_bwd(eps, res, g):
+    invvar, x, weight = res
+    grad_input, grad_weight, part_grad = _rms_norm_bwd_p.bind(
+        g, invvar, x, weight, eps=eps
+    )
+    return grad_input, grad_weight
+```
+
+`rms_norm_fwd` now returns an extra output `(invvar, x, weight)` for the residual data and `rms_norm_bwd` takes `eps`, `res`, and `g` as the parameters.
+
+Once the relationship between `rms_norm_fwd` and `rms_norm_bwd` is estabilished through `jax.custom_vjp`, JAX will ensure that the residual data from `rms_norm_fwd` is passed to `rms_norm_bwd` as `res` for backward operation.
+For non-differentiable parameters such as `eps`, JAX ensures that they are passed to the backward operation before the residual data.  That's why `eps` precedes `res` in the parameter list of `rms_norm_bwd`.
+
+Now that `rms_norm_fwd` returns the residual data, which is not needed for simple RMS normalization operation, we define a wrapper around it, `rms_norm`.  It simply calls `rms_norm_fwd` and returns only `output`.  Note that `rms_norm` is annotated with `@partial(jax.custom_vjp, nondiff_argnums=(2,))` and we are passing `rms_norm_fwd` and `rms_norm_bwd` to `rms_norm.defvjp`.  It teaches JAX that, when `rms_norm` is differentiated, `rms_norm_fwd` is to be used for forward operation, and `rms_norm_bwd` is to be used for backward operation.
+
+See [Custom derivative rules for JAX-transformable Python functions](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html#use-jax-custom-vjp-to-define-custom-reverse-mode-only-rules) for more information on `jax.custom_vjp`.
+
+```python
+@partial(jax.custom_vjp, nondiff_argnums=(2,))
+def rms_norm(x, weight, eps=1e-05):
+    output, _ = rms_norm_fwd(x, weight, eps=eps)
+    return output
+
+
+rms_norm.defvjp(rms_norm_fwd, rms_norm_bwd)
+```
+
+With the refinement we have made, the backward operation test works with a modification: `loss` now calls `rms_norm` instead of `rms_norm_fwd`.
+
+```python
+def loss(x, weight):
+    predictions = rms_norm(x, weight)
+    return -jax.numpy.mean(predictions**2)
+
+
+loss_grad = jax.grad(loss)
+out = loss_grad(x, weight)
```

Comment:
let's also test your gradient implementation with jtu.check_grad, see example

https://github.com/google/jax/blob/main/tests/experimental_rnn_test.py#L78-L79

Reply:
I haven't used it before.  let me take a look

Reply:
Done.  Added `jtu.check_grad`